### PR TITLE
Add support for Azure DevOps and BitBucket

### PR DESCRIPTION
**Pull request by csandman**
This would partially fulfill the requests from #2, however it is missing the original two sites requested, GitLab and Gitea.  I was going to do GitLab, but it appears they already offer icons from the material icon pack so I decided to leave that out for now. I also thought it best not to make this PR too large. Someone else could definitely add to this if they wanted to though.

This PR adds support for Azure DevOps and BitBucket by separating all of the site specific code and element selectors into a new object specific to each one.  I'm not sure if this new code style is formatted like you'd expect, but I'm willing to change it if need be. There were a lot of changes here, so I wouldn't be surprised if there are things you want changed. I could also split this into two separate PR's for each of the Git providers if you'd like, but keeping it all together makes it a little easier to see why I broke things down like I did. I'll also try and explain my reasoning for this breakdown here, but I also added comments to the code where it was less than obvious why I made certain choices.

---

The main part of this change was to create an object for each of the pieces of this code that are specific to each git provider's website.  These objects look like this:

```js
{
  name: 'github',
  selectors: {
    row: '.js-navigation-container[role=grid] > .js-navigation-item',
    filename: 'div[role="rowheader"] > span',
    icon: '.octicon',
  },
  getIsLightTheme: () =>
    document.querySelector('html').getAttribute('data-color-mode') === 'light',
  getIsDirectory: (svgEl) => svgEl.getAttribute('aria-label') === 'Directory',
  getIsSubmodule: (svgEl) => svgEl.getAttribute('aria-label') === 'Submodule',
  getIsSymlink: (svgEl) => svgEl.getAttribute('aria-label') === 'Symlink Directory',
  replaceIcon: (svgEl, newSVG) => {
    svgEl
      .getAttributeNames()
      .forEach((attr) => newSVG.setAttribute(attr, svgEl.getAttribute(attr)));

    svgEl.parentNode.replaceChild(newSVG, svgEl);
  },
}
```

This allowed me to keep most of the actual code for selecting the icon in the common loop of the code instead of separating it out, except for the `replaceIcon` part which is a bit longer. That function needed to be independent as all of the git providers had different structures for the element, and directly running the same original code for each icon left things styled very inconsistently.  For example, BitBucket's icon styles were being set using an `svg` selector, which wouldn't work on the new icon because an `img` element is used instead.

BitBucket works very similarly to the way GitHub works, but Azure DevOps raised some issues because of the nature of their repository file tree.

First of all, replacing the original icon element with the new icon completely broke the site every time you navigated to a new folder in a repository.  So alternatively, I added the icon as a child of the previous icon element and added styling to hide their original icon. There could be a better approach but this seemed like the safest way to handle it.

Second of all, on the first render the elements are all replaced fine, but when you expanded one of the file tree leaves, the `observe` function wasn't getting properly triggered for the new elements, and it was leaving old icons attached to new rows.  It's hard to explain exactly what was happening but it was essentially just broken.  Here's an example of what it looked like:

image

So my solution was to add a native `MutationObserver` to each of the row elements in addition to the basic observe function, but limiting the icon updates to only mutations that were not triggered by the extension itself, otherwise it would end up in an infinite loop of replacing the icons.

I also changed the way light/dark mode is detected, as the user's light/dark mode system preference might be different from their selected preference on each site.  For GitHub it was easy enough as they add an attribute to the `html` which explicitly says which mode is selected.  Azure does not have something as straightforward as this so I added a check for the body's font color.  This should be enough, considering they don't have a bunch of alternative themes (only light or dark) unless they decide to change the font color in the future.

---

If you'd like to test this, you can easily find open source projects on BitBucket to look through, like this one: https://bitbucket.org/galad87/subler/src/master/

For Azure however, you'll need to create an account and make your own repository to test with because (as far as I'm aware) there is no public setting for repos there. It is free to get started though.

Here are some screenshots of the end result for each of them:

### Bitbucket

image

image

### Azure DevOps

image

image

image

### CODE REVIEW

File: src/main.js [hunk 1]
```diff
@@ -40,10 +40,144 @@ const rushFirst = (rushBatch, callback) => {
...
+
+/**
+ * Get all selectors and functions specific to the Git provider
+ *
+ * @returns {object} All of the values needed for the provider
+ */
+const getGitProvider = () => {
+  const hostname = window.location.hostname;
+
+  if (/.*github.*/.test(hostname)) {
```
Comment (jamiehaywood) [inline src/main.js:54]: this is more of a stylistic PR comment than anything functional:

I would split this into a switch:

```js
switch (hostname) {
  case /.*github.*/.test(hostname):
    return providerConfig.github;
    break;
  case /.*bitbucket.*/.test(hostname):
    return providerConfig.bitbucket;
    break;
  case /.*azure.*/.test(hostname):
    return providerConfig.azure;
    break;

  default:
    return {};
    break;
}
```

and have the config defined in another file:

// provider.config.js
```js
const providerConfig = {
    github: {
      name: 'github',
      selectors: {
        row: '.js-navigation-container[role=grid] > .js-navigation-item',
        filename: 'div[role="rowheader"] > span',
        icon: '.octicon',
      },
      getIsLightTheme: () =>
        document.querySelector('html').getAttribute('data-color-mode') === 'light',
      getIsDirectory: (svgEl) => svgEl.getAttribute('aria-label') === 'Directory',
      getIsSubmodule: (svgEl) => svgEl.getAttribute('aria-label') === 'Submodule',
      getIsSymlink: (svgEl) => svgEl.getAttribute('aria-label') === 'Symlink Directory',
      replaceIcon: (svgEl, newSVG) => {
        svgEl
          .getAttributeNames()
          .forEach((attr) => newSVG.setAttribute(attr, svgEl.getAttribute(attr)));

        svgEl.parentNode.replaceChild(newSVG, svgEl);
      },
    },
   // bitbucket
   // azure
   // other source providers
}
```

**Review by jamiehaywood — APPROVED** [review]
bloody great - one stylistic comment, but I've been using a build of this extension for a couple of months, and it works like a dream. really great job, and thank you!!!

File: src/main.js (hunk 1 above)
Comment (jamiehaywood) [reply src/main.js:54]: or if you don't want to use a switch, you could do an object literal lookup with a regex

Comment (csandman) [reply src/main.js]: Just pushed a commit with this change!

**Comment by csandman** [comment]
Glad you like it!  I pushed a new commit with the change requested (however I removed the break statements as they all follow return statements and are unnecessary).  Let me know if there is anything else you need changed!

**Comment by jamiehaywood** [comment]
I'm unfortunately not the/a maintainer but I thought I would give my two cents anyway, as I think this should be merged and pushed

**Comment by csandman** [comment]
Ah, my bad I totally missed that haha.  Well, if @Claudiohbsantos checks it out at some point, it would be great to see this merged!

**Comment by Claudiohbsantos** [comment]
I'm really sorry! No idea how I've missed this for so long 🤦. I recently moved to Gitlab at work so my GitHub notifications have been sorely neglected.

I'll try to take a look at this and any other pending issues in this repo this weekend

File: src/main.js [hunk 2]
```diff
@@ -40,10 +42,74 @@ const rushFirst = (rushBatch, callback) => {
...
+      return providerConfig.github;
+
+    case /.*bitbucket.*/.test(hostname):
+      return providerConfig.bitbucket;
+
+    case /.*azure.*/.test(hostname):
+      return providerConfig.azure;
+
+    default:
+      return {};
```
Comment (Claudiohbsantos) [inline src/main.js:67]: I think it's pretty dangerous to return an empty object here. (well, not really dangerous since this extension exploding doesn't really cause any problems besides the lack of icons, but you get the idea lol)

The rest of the code assumes that the fields `selectors`, `getIsDirectory`, etc are always available so if `provider` is an empty object lookups such as `provider.selectors.row` explode with a TypeError.

I'd suggest building a "no-op" provider with empty selectors and functions that don't do anything. This way provider always has an expected shape.

Another option is to return `null` and then only call `observe(provider.selectors.row, ...)` if `provider` exists. This is simpler and works for us since all uses of provider currently come from that codepath. The only reason I suggested the no-op provider is because `replaceIcon`  uses the top-level `provider` directly, so we could end up calling it at some later point from a different codepath and not realize that we need to check if the `provider` const in this module exists, which is a bit confusing. The other solution to this problem is to pass `provider` as a parameter to `replaceIcon`

In short, I would be ok with either solution here:
1. return a default 'no-op' provider when hostname is not recognized
2. return `null` and check that provider exists before calling `observe`. In this case `replaceIcon` should receive `provider` as a parameter.

2 is probably fast to implement but I'll leave it up to you to decide which you prefer if you decide to take care of this

File: src/main.js [hunk 3]
```diff
@@ -40,10 +42,74 @@ const rushFirst = (rushBatch, callback) => {
   }
 };
 
+/** The name of the class used to hide the pseudo element `:before` on Azure */
+const HIDE_PSEUDO_CLASS = 'hide-pseudo';
```
Comment (Claudiohbsantos) [inline src/main.js:46]: I think this belongs with the azure provider definition since it only related to azure  (though your comment already makes this clear, thanks!)

File: src/main.js [hunk 4]
```diff
@@ -40,10 +42,74 @@ const rushFirst = (rushBatch, callback) => {
...
+      if (!hasAddedAzureStyle) {
+        // Azure requires the icon element to be left on the page so add a style rule to hide its icon
+        document.styleSheets[0].insertRule(
+          `.${HIDE_PSEUDO_CLASS}::before { display: none !important }`,
+          0
+        );
+        // but only add the style once
+        hasAddedAzureStyle = true;
+      }
+    }
```
Comment (Claudiohbsantos) [inline src/main.js:112]: This is definitely a nitpick so feel free to ignore if it's not worth your time, but I think this doesn't belong in `main.js` since it relates only to azure. I think each provider could have a `onAdd` or `onAddition` method that is executed here. This block would be the implementation of this hook for the azure provider and the other providers would have no-op functions since they don't need the hook.

File: src/provider.config.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
+const providerConfig = {
```
Comment (Claudiohbsantos) [inline src/provider.config.js:1]: This another nitpick that is more a suggestion than a request: I love that these ended in a different file outside of `main.js` but I would take it another step further: 1 file per provider. Each of these  provider objects doesn't really care about the contents of the others so I think having a single provider per file would make it easier to read/work with a specific provider. Particularly if these providers start to grow (i.e. the other note I left about a potential `onAdd` hook) I think that could make things easier.

In this case we could go with a `providers directory` with 1 file per provider and an `index` file where all individual providers are indexes into the `providerConfig` object:

```
// on providers/index.js
import githubConfig from './github'
...

const providerConfig = {
    github: githubConfig,
    ...
}
export default providerConfig
```

File: src/main.js (hunk 3 above)
Comment (Claudiohbsantos) [reply src/main.js:46]: This depends on the creation of a `onAdd` hook as described in the other comment though so that should be considered first

**Review by Claudiohbsantos — CHANGES_REQUESTED** [review]
@csandman This PR is amazing, thanks for the detailed description and very clean code! It was very easy to review and understand the changes even after months without looking at this code properly. I specially appreciate the screenshots and the link to the bitbucket repo for tests.

There is 1 blocking bug that we need to take care of before merging (switch statement never matches. see comments). The other notes I left are all suggestions that I think would improve the general organization of the code but that I'm not considering merge-blocking so feel free to skip them if you're pressed for time.

File: src/main.js (hunk 1 above)
Comment (Claudiohbsantos) [reply src/main.js]: Thanks for reviewing this @jamiehaywood ! I agree with both suggestions!

**Comment by Claudiohbsantos** [comment]
I resolved the merge conflict I just created with my last push

File: src/main.js (hunk 3 above)
Comment (csandman) [reply src/main.js]: I moved this into the `providers/azure.js` file

File: src/main.js [hunk 6]
```diff
@@ -40,10 +42,74 @@ const rushFirst = (rushBatch, callback) => {
...
+
+/**
+ * Get all selectors and functions specific to the Git provider
+ *
+ * @returns {object} All of the values needed for the provider
+ */
+const getGitProvider = () => {
+  const hostname = window.location.hostname;
+
+  switch (hostname) {
```
[the remark this answers is not in the document]
Comment (csandman) [reply src/main.js:53]: I ended up going with the second option, using the actual names, because switching on `true` feels kind of... wrong to me.  However, this matching might need to change at some point in the future, especially if you want to make this extension function for enterprise versions of github, as mentioned in #36.

File: src/main.js (hunk 4 above)
Comment (csandman) [reply src/main.js]: I ended up adding an `onAdd` function to each.  This allowed me to put the `hasAddedAzureStyle` variable at the top level in the `providers/azure.js` file, along with the `HIDE_PSEUDO_CLASS` variable mentioned above.  This function takes two arguments, `row` and `callback`, and allows the separation of this functionality.

File: src/main.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (csandman) [reply src/main.js:53]: Also I'm not sure how I missed this bug!  I think I either ended up forgetting to build before testing these changes or I assumed they would work because I thought it was you who made the comments about the change haha (my bad).  Thanks for catching this!

**Comment by csandman** [comment]
I tackled all of the comments you left on this PR, hopefully everything is up to snuff now!  Sorry about that bug, I somehow missed that when making the changes recommended by @jamiehaywood.

**Comment by csandman** [comment]
Sorry, I accidentally merged my eslint branch into this one, not really sure how that happened. I meant to do it the other way around haha.

File: src/main.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (csandman) [reply src/main.js]: Scratch what I said before, I gave it some more though, and I think the first option here will allow more flexibility in the future. I switched back to a `.test` on the URL to allow for something like the GitHub enterprise URL's in the future, which would probably need a pattern to match instead of a fixed URL.  I also switched to testing the full `href` for further flexibility.  LMK if this is alright!

**Comment by csandman** [comment]
I ended up adding one last change to the core functionality of this PR that I hope isn't outside of it's scope.  I ended up changing the selectors for the GitHub config to allow for replacement in the new PR file explorer view.

image

I originally tried making a new provider config that matched that URL but I realized it wouldn't work for the normal view if you navigated between pages. So instead I just modified the selectors to allow it to select either the rows in the standard codebase view or the file tree view.  I also had to make the icon selector more specific to avoid matching the chevron icons in the file tree view.

Overall, I can't think of any scenario where having these OR css selectors would cause any conflicts.

If this is outside of the scope of this PR and you want me to revert, lmk.

You can see the change here: https://github.com/Claudiohbsantos/github-material-icons-extension/pull/31/commits/972c5e186ca03b0a6ea68175917c43202a65f339

File: src/main.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Claudiohbsantos) [reply src/main.js]: This is perfect, thanks!

File: src/main.js (hunk 4 above)
Comment (Claudiohbsantos) [reply src/main.js]: That looks great, thanks for moving it!

**Review by Claudiohbsantos — APPROVED** [review]
Thanks for tackling the notes so quickly, this looks great! And the file tree icons look amazing, great addition! I can't think of a reason not to use the selectors as you did, let's leave it like that and revisit if there are any issues.

I'm going to merge this into master and aim for a release this week.
